### ISSUE TITLE
Specifying one character as a name of a volume shows an unexpected result

### ISSUE THREAD

**Issue by ozlerhakan** [docker/docker#20122]
Hi,

I need to give a new to each managed volume in alphabetic order, but it seems like this shows unexpected result in the file system of the container. I don't see the archive folder, instead I see `a:` and `b:`

How reproducible:

```
$ docker run --name ubuntu -v a:/archive ubuntu:14.04 ls -lha
total 76K
drwxr-xr-x  33 root root 4.0K Feb  8 22:30 .
drwxr-xr-x  33 root root 4.0K Feb  8 22:30 ..
-rwxr-xr-x   1 root root    0 Feb  8 22:30 .dockerenv
-rwxr-xr-x   1 root root    0 Feb  8 22:30 .dockerinit
drwxr-xr-x   3 root root 4.0K Feb  8 22:30 a:    <<<----
drwxr-xr-x   2 root root 4.0K Jan 19 12:50 bin
.....

$ docker run --name ubuntu2 -v b:/archive ubuntu:14.04 ls -lha
total 76K
drwxr-xr-x  33 root root 4.0K Feb  8 22:31 .
drwxr-xr-x  33 root root 4.0K Feb  8 22:31 ..
-rwxr-xr-x   1 root root    0 Feb  8 22:31 .dockerenv
-rwxr-xr-x   1 root root    0 Feb  8 22:31 .dockerinit
drwxr-xr-x   3 root root 4.0K Feb  8 22:31 b: <<<----
drwxr-xr-x   2 root root 4.0K Jan 19 12:50 bin
....

$ docker volume ls
DRIVER              VOLUME NAME
local               8e19cd41b34ecb2558a08ccdf27381562919373420df6d3d16adf054b6185e32
local               6c8823e16e8f64465fed79cf9c727e5291a5705feb77de23c91c489b5eb8dd30
```

But simply adding 2 characters as a name works just fine:
```
$ docker run --name ubuntu -v ab:/archive ubuntu:14.04 ls -lha
total 76K
drwxr-xr-x  33 root root 4.0K Feb  8 22:33 .
drwxr-xr-x  33 root root 4.0K Feb  8 22:33 ..
-rwxr-xr-x   1 root root    0 Feb  8 22:33 .dockerenv
-rwxr-xr-x   1 root root    0 Feb  8 22:33 .dockerinit
drwxr-xr-x   2 root root 4.0K Feb  8 22:33 archive  <<<-- here is the file!
.....

$ docker run --name ubuntu2 -v cd:/archive ubuntu:14.04 ls -lha
total 76K
drwxr-xr-x  33 root root 4.0K Feb  8 22:34 .
drwxr-xr-x  33 root root 4.0K Feb  8 22:34 ..
-rwxr-xr-x   1 root root    0 Feb  8 22:34 .dockerenv
-rwxr-xr-x   1 root root    0 Feb  8 22:34 .dockerinit
drwxr-xr-x   2 root root 4.0K Feb  8 22:34 archive   <<<-- and here it is
drwxr-xr-x   2 root root 4.0K Jan 19 12:50 bin
drwxr-xr-x   2 root root 4.0K Apr 10  2014 boot
..............

$ docker volume ls
DRIVER              VOLUME NAME
local               ab
local               cd

```

`docker version`:
```
Client:
 Version:      1.10.0
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   590d5108
 Built:        Thu Feb  4 18:18:11 2016
 OS/Arch:      windows/amd64

Server:
 Version:      1.10.0
 API version:  1.22
 Go version:   go1.5.3
 Git commit:   590d5108
 Built:        Thu Feb  4 19:55:25 2016
 OS/Arch:      linux/amd64
```

`docker info`:
```
Containers: 2
 Running: 0
 Paused: 0
 Stopped: 2
Images: 1
Server Version: 1.10.0
Storage Driver: aufs
 Root Dir: /mnt/sda1/var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 122
 Dirperm1 Supported: true
Execution Driver: native-0.2
Logging Driver: json-file
Plugins:
 Volume: local
 Network: host bridge null
Kernel Version: 4.1.17-boot2docker
Operating System: Boot2Docker 1.10.0 (TCL 6.4.1); master : b09ed60 - Thu Feb  4 20:16:08 UTC 2016
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 996.2 MiB
Name: kodcu
ID: 32KK:K7BX:D2KV:SXYC:N65U:2FBX:IUQD:U24R:NBYR:EEV4:BBPU:Y66D
Debug mode (server): true
 File Descriptors: 10
 Goroutines: 22
 System Time: 2016-02-07T14:20:50.732065775Z
 EventsListeners: 0
 Init SHA1:
 Init Path: /usr/local/bin/docker
 Docker Root Dir: /mnt/sda1/var/lib/docker
Username: 
Registry: https://index.docker.io/v1/
Labels:
 provider=virtualbox
```

**Comment by cpuguy83** [comment]
lol, this is because of windows support.
Short and quick fix, don't do that.

ping @tiborvass WDYT?

**Comment by thaJeztah** [comment]
Don't we disallow single-character names? On Windows we could make a trailing slash/backslash required

**Comment by thaJeztah** [comment]
(For single character names)

**Comment by cpuguy83** [comment]
@thaJeztah No, we don't disallow single char volume names, and even if we did, this is getting caught before we hit the naming aspect since the client is parsing these names first (IIRC).

This is the complex parsing logic that @tiborvass made so that we can support both Windows and Linux daemons from either a Linux client or a Windows client w/o changing the API.

**Comment by thaJeztah** [comment]
if we can't fix it without adding a lot of complexity, then I agree we should close this

**Comment by tiborvass** [comment]
> a:/archive
> b:/archive

:floppy_disk: :joy: 

@cpuguy83 More seriously, I don't have any objections of restricting volume names to 2 or more characters.

**Comment by ozlerhakan** [comment]
I just realize that, `$docker volume create`  doesn't let me create a single char name 

```
$ docker volume create --name a
Error response from daemon: create a: volume name invalid: "a" includes invalid characters for a local volume name, only "[a-zA-Z0-9][a-zA-Z0-9_.-]" are allowed
```

**Comment by jglick** [comment]
This is a regression in the 1.10.0 client which makes the Jenkins Docker Pipeline plugin not work with projects named with a single letter. It is not just abstract volume names which are affected, but absolute file paths that happen to end in a one-letter directory. Can we at least disable the hack when the host dir is an absolute path, which ought to be unambiguous?

```
$ for v in 1.7.1 1.8.0 1.8.1 1.8.2 1.8.3 1.9.0 1.9.1 1.10.0 1.10.1; do echo $v; curl -s https://get.docker.com/builds/Linux/x86_64/docker-$v > docker; chmod a+x docker; for n in xxx xx x; do ./docker run -v /tmp/$n:/tmp/$n:rw ubuntu sh -c 'mount|fgrep /tmp/x'; done; done
1.7.1
/dev/dm-1 on /tmp/xxx type ext4 (rw,noatime,nodiratime,errors=remount-ro,data=ordered)
/dev/dm-1 on /tmp/xx type ext4 (rw,noatime,nodiratime,errors=remount-ro,data=ordered)
/dev/dm-1 on /tmp/x type ext4 (rw,noatime,nodiratime,errors=remount-ro,data=ordered)
1.8.0
/dev/dm-1 on /tmp/xxx type ext4 (rw,noatime,nodiratime,errors=remount-ro,data=ordered)
/dev/dm-1 on /tmp/xx type ext4 (rw,noatime,nodiratime,errors=remount-ro,data=ordered)
/dev/dm-1 on /tmp/x type ext4 (rw,noatime,nodiratime,errors=remount-ro,data=ordered)
1.8.1
/dev/dm-1 on /tmp/xxx type ext4 (rw,noatime,nodiratime,errors=remount-ro,data=ordered)
/dev/dm-1 on /tmp/xx type ext4 (rw,noatime,nodiratime,errors=remount-ro,data=ordered)
/dev/dm-1 on /tmp/x type ext4 (rw,noatime,nodiratime,errors=remount-ro,data=ordered)
1.8.2
/dev/dm-1 on /tmp/xxx type ext4 (rw,noatime,nodiratime,errors=remount-ro,data=ordered)
/dev/dm-1 on /tmp/xx type ext4 (rw,noatime,nodiratime,errors=remount-ro,data=ordered)
/dev/dm-1 on /tmp/x type ext4 (rw,noatime,nodiratime,errors=remount-ro,data=ordered)
1.8.3
/dev/dm-1 on /tmp/xxx type ext4 (rw,noatime,nodiratime,errors=remount-ro,data=ordered)
/dev/dm-1 on /tmp/xx type ext4 (rw,noatime,nodiratime,errors=remount-ro,data=ordered)
/dev/dm-1 on /tmp/x type ext4 (rw,noatime,nodiratime,errors=remount-ro,data=ordered)
1.9.0
/dev/dm-1 on /tmp/xxx type ext4 (rw,noatime,nodiratime,errors=remount-ro,data=ordered)
/dev/dm-1 on /tmp/xx type ext4 (rw,noatime,nodiratime,errors=remount-ro,data=ordered)
/dev/dm-1 on /tmp/x type ext4 (rw,noatime,nodiratime,errors=remount-ro,data=ordered)
1.9.1
/dev/dm-1 on /tmp/xxx type ext4 (rw,noatime,nodiratime,errors=remount-ro,data=ordered)
/dev/dm-1 on /tmp/xx type ext4 (rw,noatime,nodiratime,errors=remount-ro,data=ordered)
/dev/dm-1 on /tmp/x type ext4 (rw,noatime,nodiratime,errors=remount-ro,data=ordered)
1.10.0
/dev/dm-1 on /tmp/xxx type ext4 (rw,noatime,nodiratime,errors=remount-ro,data=ordered)
/dev/dm-1 on /tmp/xx type ext4 (rw,noatime,nodiratime,errors=remount-ro,data=ordered)
/dev/dm-1 on /tmp/x:/tmp/x:rw type ext4 (rw,noatime,nodiratime,errors=remount-ro,data=ordered)
1.10.1
/dev/dm-1 on /tmp/xxx type ext4 (rw,noatime,nodiratime,errors=remount-ro,data=ordered)
/dev/dm-1 on /tmp/xx type ext4 (rw,noatime,nodiratime,errors=remount-ro,data=ordered)
/dev/dm-1 on /tmp/x:/tmp/x:rw type ext4 (rw,noatime,nodiratime,errors=remount-ro,data=ordered)
```

**Comment by aidanhs** [comment]
> It is not just abstract volume names which are affected, but absolute file paths that happen to end in a one-letter directory.

Exactly this, I regularly do `docker run -it --rm -v $(pwd):/x ubuntu:14.04.3 bash` just to jump in a container to play with something and discovered to my utter confusion today that if I'm in a directory with a single letter name, volumes silently don't work! (note that I've altered the command to clarify that the `:/x` isn't the problem)

```
/tmp/tmp.PPQGhNXCWt/x $ docker run -it --rm -v $(pwd):/abcdef ubuntu:14.04.3 bash
root@fae54c55ffe0:/# ls
bin  boot  dev  etc  home  lib  lib64  media  mnt  opt  proc  root  run  sbin  srv  sys  tmp  usr  var
```

**Comment by aidanhs** [comment]
@thaJeztah surely this is at least partially a bug rather than a feature? "You can't mount directories with a single character name" is pretty unexpected!

**Comment by cpuguy83** [comment]
@aidanhs yes, there's definitely a bug involved here.

**Comment by thaJeztah** [comment]
Let me add this to the 1.11 milestone and try to get it fixed in 1.11

**Comment by yongtang** [comment]
I can look into this issue and see if I can provide a bug fix soon. #dibs
